src/topology/ping_pong.rs

```diff
@@ -0,0 +1,865 @@
+    A::PrepareMessage: PartialEq,
+{
+    fn decode_with_param(
+        decoding_param: &PrepareStateDecode,
+        bytes: &mut std::io::Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        let previous_prepare_state = A::PrepareState::decode_with_param(decoding_param, bytes)?;
+        let current_prepare_message =
+            A::PrepareMessage::decode_with_param(&previous_prepare_state, bytes)?;
+
+        Ok(Self {
+            previous_prepare_state,
+            current_prepare_message,
+        })
+    }
+}
+
+/// Corresponds to the `State` enumeration implicitly defined in [VDAF's Ping-Pong Topology][VDAF].
+/// VDAF describes `Start` and `Rejected` states, but the `Start` state is never instantiated in
+/// code, and the `Rejected` state is represented as `std::result::Result::Err`, so this enum does
+/// not include those variants.
+///
+/// [VDAF]: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-06#section-5.8
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum State<
+    const VERIFY_KEY_SIZE: usize,
+    const NONCE_SIZE: usize,
+    A: Aggregator<VERIFY_KEY_SIZE, NONCE_SIZE>,
+> {
+    /// Preparation of the report will continue with the enclosed state.
+    Continued(A::PrepareState),
+    /// Preparation of the report is finished and has yielded the enclosed output share.
+    Finished(A::OutputShare),
+}
+
+/// Values returned by [`PingPongTopology::continued`].
+#[derive(Clone, Debug)]
+pub enum ContinuedValue<
+    const VERIFY_KEY_SIZE: usize,
+    const NONCE_SIZE: usize,
+    A: Aggregator<VERIFY_KEY_SIZE, NONCE_SIZE>,
+> {
+    /// The operation resulted in a new state and a message to transmit to the peer.
+    WithMessage {
+        /// The transition that will be executed. Call `Transition::evaluate` to obtain the next
+        /// [`State`] and a [`Message`] to transmit to the peer.
+        transition: Transition<VERIFY_KEY_SIZE, NONCE_SIZE, A>,
+    },
+    /// The operation caused the host to finish preparation of the input share, yielding an output
+    /// share and no message for the peer.
+    FinishedNoMessage {
+        /// The output share which may now be accumulated.
+        output_share: A::OutputShare,
+    },
+}
+
+/// Extension trait on [`crate::vdaf::Aggregator`] which adds the [VDAF Ping-Pong Topology][VDAF].
+///
+/// [VDAF]: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-06#section-5.8
+pub trait PingPongTopology<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize>:
+    Aggregator<VERIFY_KEY_SIZE, NONCE_SIZE>
+{
+    /// Specialization of [`State`] for this VDAF.
+    type State;
+    /// Specialization of [`ContinuedValue`] for this VDAF.
+    type ContinuedValue;
+    /// Specializaton of [`Transition`] for this VDAF.
+    type Transition;
+
+    /// Initialize leader state using the leader's input share. Corresponds to
+    /// `ping_pong_leader_init` in the forthcoming `draft-irtf-cfrg-vdaf-07`.
```

Comment:
nit: For consistency, maybe
```suggestion
    /// `ping_pong_leader_init` in [VDAF]`.
```
and add the link below. Also, since we're cutting 07 imminently, drop "forthcoming".

Here and below.